agent/agent_test.go

```diff
@@ -91,71 +92,112 @@ func (s *AgentSuite) TestNextTaskResponseShouldExit() {
 		TaskId:     "mocktaskid",
 		TaskSecret: "",
 		ShouldExit: true}
-	ctx, cancel := context.WithCancel(context.Background())
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 
-	err := s.a.loop(ctx)
-	s.NoError(err)
+	errs := make(chan error, 1)
+	go func() {
+		errs <- s.a.loop(ctx)
+	}()
+	select {
+	case err := <-errs:
+		s.NoError(err)
+	case <-ctx.Done():
+		s.FailNow(ctx.Err().Error())
+	}
 }
 
 func (s *AgentSuite) TestTaskWithoutSecret() {
 	s.mockCommunicator.NextTaskResponse = &apimodels.NextTaskResponse{
 		TaskId:     "mocktaskid",
 		TaskSecret: "",
 		ShouldExit: false}
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 
-	errChan := make(chan error)
+	agentCtx, agentCancel := context.WithCancel(ctx)
+	errs := make(chan error, 1)
 	go func() {
-		errChan <- s.a.loop(ctx)
+		errs <- s.a.loop(agentCtx)
 	}()
 	time.Sleep(1 * time.Second)
-	cancel()
-	err := <-errChan
-	s.NoError(err)
+	agentCancel()
+	select {
+	case err := <-errs:
+		s.NoError(err)
+	case <-ctx.Done():
+		s.FailNow(ctx.Err().Error())
+	}
 }
 
 func (s *AgentSuite) TestErrorGettingNextTask() {
 	s.mockCommunicator.NextTaskShouldFail = true
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 
-	err := s.a.loop(ctx)
-	s.Error(err)
+	errs := make(chan error, 1)
+	go func() {
+		errs <- s.a.loop(ctx)
+	}()
+	select {
+	case err := <-errs:
+		s.Error(err)
+	case <-ctx.Done():
+		s.FailNow(ctx.Err().Error())
+	}
 }
 
 func (s *AgentSuite) TestCanceledContext() {
 	s.a.opts.AgentSleepInterval = time.Millisecond
 	s.a.opts.MaxAgentSleepInterval = time.Millisecond
 	s.mockCommunicator.NextTaskIsNil = true
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
I pass the agentCtx below and cancel it. I use this timeout just to put a reasonable deadline on how long the test can run, because that if the agent loop has a bug and/or panics, the test could just just be stuck looping until it hits the 10 minute go test timeout.